File: q2353/q2353.py
class Food: 
    def __init__(self, name, rating, cuisine):
        self.name = name
        self.rating = rating
        self.cuisine = cuisine
    
    def __lt__(self, other):
        if self.rating == other.rating: 
            return self.name < other.name
        return self.rating > other.rating
    
class FoodRatings:

    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.foodMap = {}
        self.cuisineFoodMap = defaultdict(list)
        for i in range(len(foods)):
            name, rating, cuisine = foods[i], ratings[i], cuisines[i]
            self.foodMap[name] = Food(name, rating, cuisine)
            heapq.heappush(self.cuisineFoodMap[cuisine], Food(name, rating, cuisine))

    def changeRating(self, food: str, newRating: int) -> None:
        self.foodMap[food].rating = newRating
        cuisine = self.foodMap[food].cuisine
        heapq.heappush(self.cuisineFoodMap[cuisine], Food(food, newRating, cuisine))

    def highestRated(self, cuisine: str) -> str:
        highestRatedFood = self.cuisineFoodMap[cuisine][0]
        while (highestRatedFood.rating != self.foodMap[highestRatedFood.name].rating):
            heapq.heappop(self.cuisineFoodMap[cuisine])
            highestRatedFood = self.cuisineFoodMap[cuisine][0]
        return highestRatedFood.name

File: q2353/q2353.cpp
class Food
{
public:
    string name;
    int rating;
    string cuisine;

    Food(string name, int rating, string cuisine)
    {
        this->name = name;
        this->rating = rating;
        this->cuisine = cuisine;
    }

    bool operator<(const Food &other) const
    {
        if (this->rating == other.rating)
        {
            return this->name > other.name;
        }
        return this->rating < other.rating;
    }
};

class FoodRatings
{
public:
    FoodRatings(vector<string> &foods, vector<string> &cuisines, vector<int> &ratings)
    {
        for (int i = 0; i < foods.size(); i++)
        {
            string name = foods[i];
            int rating = ratings[i];
            string cuisine = cuisines[i];
            foodMap.insert(make_pair(name, new Food(name, rating, cuisine)));
            cuisineFoodMap[cuisine].push(Food(name, rating, cuisine));
        }
    }

    void changeRating(string food, int newRating)
    {
        auto foodItem = foodMap[food];
        foodItem->rating = newRating;
        string cuisine = foodMap[food]->cuisine;
        cuisineFoodMap[cuisine].push(Food(food, newRating, cuisine));
    }

    string highestRated(string cuisine)
    {
        auto highestRatedFood = cuisineFoodMap[cuisine].top();
        while (highestRatedFood.rating !=
               foodMap[highestRatedFood.name]->rating)
        {
            cuisineFoodMap[cuisine].pop();
            highestRatedFood = cuisineFoodMap[cuisine].top();
        }
        return highestRatedFood.name;
    }

private:
    unordered_map<string, shared_ptr<Food>> foodMap;
    unordered_map<string, priority_queue<Food>> cuisineFoodMap;
};